clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -1985,47 +1985,63 @@ inlineHO _ e@(App _ _)
 inlineHO _ e = return e
 {-# SCC inlineHO #-}
 
--- | Simplified CSE, only works on let-bindings, works from top to bottom
+-- | Simplified CSE, only works on let-bindings, does an inverse topological
+-- sort of the let-bindings and then works from top to bottom
+--
+-- XXX: Check whether inverse top-sort followed by single traversal removes as
+-- many binders as the previous "apply-until-fixpoint" approach in the presence
+-- of recursive groups in the let-bindings. If not but just for checking whether
+-- changes to transformation affect the eventual size of the circuit, it would
+-- be really helpful if we tracked circuit size in the regression/test suite.
+-- On the two examples that were tested, Reducer and PipelinesViaFolds, this new
+-- version of CSE removed the same amount of let-binders.
 simpleCSE :: HasCallStack => NormRewrite
-simpleCSE (TransformContext is0 _) e@(Letrec binders body) = do
-  let is1 = extendInScopeSetList is0 (map fst binders)
-  let (reducedBindings,body') = reduceBindersFix is1 binders body
-  if length binders /= length reducedBindings
-     then changed (Letrec reducedBindings body')
-     else return e
+simpleCSE (TransformContext is0 _) (inverseTopSortLetBindings -> Letrec bndrs body) = do
+  let is1 = extendInScopeSetList is0 (map fst bndrs)
+  (subst,bndrs1) <- reduceBinders (mkSubst is1) [] bndrs
+  -- TODO: check whether a substitution over the body is enough, the reason I'm
+  -- doing a substitution over the the binders as well is that I don't know in
+  -- what order a recursive group shows up in a inverse topological sort.
+  -- Depending on the order and forgetting to apply the substitution over the
+  -- let-bindings might lead to the introduction of free variables.
+  --
+  -- NB: don't apply the substitution to the entire let-expression, and that
+  -- would rename the let-bindings because they've been added to the InScopeSet
+  -- of the substitution.
+  let bndrs2 = map (second (substTm "simpleCSE.bndrs" subst)) bndrs1
+      body1  = substTm "simpleCSE.body" subst body
+  return (Letrec bndrs2 body1)
 
 simpleCSE _ e = return e
 {-# SCC simpleCSE #-}
 
-reduceBindersFix
-  :: InScopeSet
-  -> [LetBinding]
-  -> Term
-  -> ([LetBinding],Term)
-reduceBindersFix is binders body =
-  if length binders /= length reduced
-     then reduceBindersFix is reduced body'
-     else (binders,body)
- where
-  (reduced,body') = reduceBinders is [] body binders
-
+-- | XXX: is given inverse topologically sorted binders, but returns
+-- topologically sorted binders
+--
+-- TODO: check further speed improvements:
+--
+-- 1. Store the processed binders in a `Map Expr LetBinding`:
+--    * Trades O(1) `cons` and O(n)*aeqTerm `find` for:
+--    * O(log n)*aeqTerm `insert` and O(log n)*aeqTerm `lookup`
```

Comment:
Like discussed offline: this applies to a single iteration (of the loop implemented in `reduceBinders`). Because we're doing it approx `½n` times, these complexities become `O(n)`/`O(n2)*aeqTerm` and `O(n log n)*aeqTerm`/ `O(n log n)*aeqTerm`.

Reply:
I'm not sure I see what you're getting at with this comment @martijnbastiaan. Can you elaborate?